File: third-party/thrift/src/thrift/lib/rust/src/dep_tests/binary.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use anyhow::Result;
use fbthrift::binary_protocol::deserialize;
use fbthrift::binary_protocol::serialize;
use fbthrift::ttype::TType;
use fbthrift_test_if::Un;
use proptest::prelude::*;

use crate::proptest::gen_main_struct;

#[test]
fn test_unknown_union() -> Result<()> {
    // Build the empty union
    let u = Un::default();

    let s = serialize(&u);
    // only TType::Stop
    assert_eq!(&[TType::Stop as u8], s.as_ref());

    // Assert that deserialize builds the exact some struct
    assert_eq!(u, deserialize(s).unwrap());

    // ...
    // extra weirdness
    // Build an explicit unknown
    let explicit_unknown = Un::UnknownField(100);
    let s2 = serialize(&explicit_unknown);
    // only Stop
    assert_eq!(&[TType::Stop as u8], s2.as_ref());

    Ok(())
}

proptest! {
#[test]
fn test_prop_serialize_deserialize(s in gen_main_struct()) {
    let processed = deserialize(serialize(&s)).unwrap();
    prop_assert_eq!(s, processed);
}
}

File: third-party/thrift/src/thrift/lib/cpp2/dynamic/Binary.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <thrift/lib/cpp2/dynamic/detail/Datum.h>

#include <folly/io/Cursor.h>
#include <folly/io/IOBuf.h>

namespace apache::thrift::dynamic {

#if FOLLY_HAVE_MEMORY_RESOURCE

Binary::Binary(const Binary& other) : Binary(other.clone(other.mr_)) {}

Binary& Binary::operator=(const Binary& other) {
  if (this != &other) {
    *this = other.clone(other.mr_);
  }
  return *this;
}

Binary Binary::clone(std::pmr::memory_resource* mr) const {
  if (!data_) {
    return Binary(mr);
  }
  return Binary(data_->clone(mr));
}

#else

Binary::Binary(const Binary& other) : Binary(other.clone()) {}

Binary& Binary::operator=(const Binary& other) {
  if (this != &other) {
    *this = other.clone();
  }
  return *this;
}

Binary Binary::clone(std::pmr::memory_resource* mr) const {
  if (!data_) {
    return Binary(mr);
  }
  return Binary(data_->clone());
}

#endif // FOLLY_HAVE_MEMORY_RESOURCE

folly::io::Cursor Binary::cursor() const {
  return folly::io::Cursor(data_.get());
}

bool operator==(const Binary& lhs, const Binary& rhs) noexcept {
  // If both are null, they're equal
  if (!lhs.data_ && !rhs.data_) {
    return true;
  }
  // If only one is null, they're not equal
  if (!lhs.data_ || !rhs.data_) {
    return false;
  }
  // Compare IOBuf contents
  return folly::IOBufEqualTo()(*lhs.data_, *rhs.data_);
}

} // namespace apache::thrift::dynamic
